python/az/aro/azext_aro/custom.py

```diff
@@ -155,6 +160,116 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
+
+        rp_client_sp_id = aad.get_service_principal_id(resolve_rp_client_id())
+        if not rp_client_sp_id:
+            if tempAppCreated:
+                aad.delete_application(client_id)
+            raise ResourceNotFoundError("RP service principal not found.")
+
+        sp_obj_ids = [client_sp_id, rp_client_sp_id]
+        cluster = mockoc(disk_encryption_set, master_subnet, worker_subnet)
+        try:
+            # Get cluster resources we need to assign permissions on, sort to ensure the same order of operations
+            resources = {ROLE_NETWORK_CONTRIBUTOR: sorted(get_cluster_network_resources(cmd.cli_ctx, cluster, True)),
+                         ROLE_READER: sorted(get_disk_encryption_resources(cluster))}
+        except (CloudError, HttpOperationError) as e:
+            logger.error(e.message)
+            if tempAppCreated:
+                aad.delete_application(client_id)
+            raise
+
+        if vnet is None:
+            master_parts = parse_resource_id(master_subnet)
+            vnet = resource_id(
+                subscription=master_parts['subscription'],
+                resource_group=master_parts['resource_group'],
+                namespace='Microsoft.Network',
+                type='virtualNetworks',
+                name=master_parts['name'],
+            )
+
+        error_objects = validate_cluster_create(cmd,
+                                                client,
+                                                master_subnet,
+                                                worker_subnet,
+                                                vnet,
+                                                pod_cidr,
+                                                service_cidr,
+                                                version,
+                                                location,
+                                                resources,
+                                                sp_obj_ids)
+        errors = []
+        for error_func in error_objects:
+            namespace = collections.namedtuple("Namespace", locals().keys())(*locals().values())
+            error_obj = error_func(cmd, namespace)
+            if error_obj != []:
+                for err in error_obj:
+                    # Wrap text so tabulate returns a pretty table
+                    new_err = []
+                    for txt in err:
+                        new_err.append(textwrap.fill(txt, width=160))
+                    errors.append(new_err)
+
+        if len(errors) > 0:
+            logger.error("Issues found blocking cluster creation.\n")
+            headers = ["Type", "Name", "Error"]
+            table = tabulate(errors, headers=headers, tablefmt="grid")
+            print(f"\n{table}")
+        else:
+            print("\nNo Issues on subscription blocking cluster creation\n")
+        if tempAppCreated:
+            aad.delete_application(client_id)
+    except Exception as e:
```

Comment:
Wherever there is a `raise` statement within the `try` clause, the code already checks `if tempAppCreated` and calls `aad.delete_application()` accordingly, so the call to `aad.delete_application()` within this `except` clause will fail in those cases - the app will already have been deleted and will no longer exist.

There are different ways to fix this, but I would do it by removing all instances of:
```
if tempAppCreated:
    aad.delete_application(client_id)
```

from the `try` clause and replacing the `except` clause with a `finally` clause. That way if an exception is raised from within the `try` clause, the code will always check whether there is a temporary app that needs to be deleted, delete it if necessary, and then raise the exception.